### PR TITLE
[WIP]pipeline with linear

### PR DESCRIPTION
**Pull request by wd665544**
## Description
[WIP]pipeline with linear

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

**Comment by yxxhero** [comment]
@wd665544 Please add more test for your PR.

File: manager/rpcserver/rpcserver.go [hunk 1]
```diff
@@ -709,12 +710,17 @@ func (s *Server) GetModel(ctx context.Context, req *managerv1.GetModelRequest) (
 	if err := s.db.WithContext(ctx).First(&scheduler, req.SchedulerId).Error; err != nil {
 		return nil, status.Error(codes.Unknown, err.Error())
 	}
-
+	logger.Infof("getModel req model id is %v", req.ModelId)
```
Comment (gaius-qi) [inline manager/rpcserver/rpcserver.go:713]: Remove log.

File: manager/rpcserver/rpcserver.go [hunk 2]
```diff
@@ -709,12 +710,17 @@ func (s *Server) GetModel(ctx context.Context, req *managerv1.GetModelRequest) (
...
+	logger.Infof("getModel req model schid is %v", req.SchedulerId)
```
Comment (gaius-qi) [inline manager/rpcserver/rpcserver.go:714]: Remove log.

File: manager/rpcserver/rpcserver.go [hunk 3]
```diff
@@ -709,12 +710,17 @@ func (s *Server) GetModel(ctx context.Context, req *managerv1.GetModelRequest) (
...
 	var model types.Model
-	if err := s.rdb.Get(ctx, cache.MakeModelKey(scheduler.SchedulerClusterID, scheduler.HostName, scheduler.IP, req.ModelId)).Scan(&model); err != nil {
+	result, err := s.rdb.Get(ctx, cache.MakeModelKey(scheduler.SchedulerClusterID, scheduler.HostName, scheduler.IP, req.ModelId)).Result()
+	if err != nil {
 		return nil, status.Error(codes.Unknown, err.Error())
 	}
-
+	err = json.Unmarshal([]byte(result), &model)
```
Comment (gaius-qi) [inline manager/rpcserver/rpcserver.go:720]:
```go
if err := xxx; err != nil {
}
```

File: manager/rpcserver/rpcserver.go [hunk 4]
```diff
@@ -856,10 +859,15 @@ func (s *Server) GetModelVersion(ctx context.Context, req *managerv1.GetModelVer
 	}
 
 	var modelVersion types.ModelVersion
-	if err := s.rdb.Get(ctx, cache.MakeModelVersionKey(scheduler.SchedulerClusterID, scheduler.HostName, scheduler.IP, req.ModelId, req.VersionId)).Scan(&modelVersion); err != nil {
+	result, err := s.rdb.Get(ctx, cache.MakeModelVersionKey(scheduler.SchedulerClusterID, scheduler.HostName, scheduler.IP, req.ModelId, req.VersionId)).Result()
```
Comment (gaius-qi) [inline manager/rpcserver/rpcserver.go:862]:
```go
modelVersion, err := xxx
```

File: manager/rpcserver/rpcserver.go [hunk 5]
```diff
@@ -709,12 +710,17 @@ func (s *Server) GetModel(ctx context.Context, req *managerv1.GetModelRequest) (
 	if err := s.db.WithContext(ctx).First(&scheduler, req.SchedulerId).Error; err != nil {
 		return nil, status.Error(codes.Unknown, err.Error())
 	}
-
+	logger.Infof("getModel req model id is %v", req.ModelId)
+	logger.Infof("getModel req model schid is %v", req.SchedulerId)
 	var model types.Model
-	if err := s.rdb.Get(ctx, cache.MakeModelKey(scheduler.SchedulerClusterID, scheduler.HostName, scheduler.IP, req.ModelId)).Scan(&model); err != nil {
+	result, err := s.rdb.Get(ctx, cache.MakeModelKey(scheduler.SchedulerClusterID, scheduler.HostName, scheduler.IP, req.ModelId)).Result()
```
Comment (gaius-qi) [inline manager/rpcserver/rpcserver.go:716]:
```go
model, err := xxx
```

File: manager/rpcserver/rpcserver.go [hunk 6]
```diff
@@ -856,10 +859,15 @@ func (s *Server) GetModelVersion(ctx context.Context, req *managerv1.GetModelVer
 	}
 
 	var modelVersion types.ModelVersion
-	if err := s.rdb.Get(ctx, cache.MakeModelVersionKey(scheduler.SchedulerClusterID, scheduler.HostName, scheduler.IP, req.ModelId, req.VersionId)).Scan(&modelVersion); err != nil {
+	result, err := s.rdb.Get(ctx, cache.MakeModelVersionKey(scheduler.SchedulerClusterID, scheduler.HostName, scheduler.IP, req.ModelId, req.VersionId)).Result()
+	if err != nil {
 		return nil, status.Error(codes.Unknown, err.Error())
 	}
-
+	logger.Infof("getmodelverison get result, result is %v", result)
```
Comment (gaius-qi) [inline manager/rpcserver/rpcserver.go:866]: Remove log.

File: manager/rpcserver/rpcserver.go [hunk 7]
```diff
@@ -856,10 +859,15 @@ func (s *Server) GetModelVersion(ctx context.Context, req *managerv1.GetModelVer
...
+	err = json.Unmarshal([]byte(result), &modelVersion)
```
Comment (gaius-qi) [inline manager/rpcserver/rpcserver.go:867]:
```go
if err := xxx; err != nil {
}
```

File: pkg/pipeline/handler.go [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+
+	logger "d7y.io/dragonfly/v2/internal/dflog"
+)
+
+type StepInfra struct {
+	name    string
+	Handler HandlerFunc
+}
+
+// HandlerFunc first chan represent input
```
Comment (gaius-qi) [inline pkg/pipeline/handler.go:15]: Syntax error.

File: pkg/pipeline/pipeline.go [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+type StepConstruct func() Step
+
+type contextVal int
+
+const (
+	OutDegree contextVal = iota
+	StepName
+)
+
+func (p *Pipeline) handleDag(ctx context.Context, cancel context.CancelFunc, graph dag.DAG[StepConstruct], fstChannel chan *Request) chan *Request {
```
Comment (gaius-qi) [inline pkg/pipeline/pipeline.go:28]: handleDAG

File: pkg/pipeline/handler.go [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+import (
+	"context"
+	"sync"
+
+	logger "d7y.io/dragonfly/v2/internal/dflog"
+)
+
+type StepInfra struct {
+	name    string
+	Handler HandlerFunc
```
Comment (gaius-qi) [inline pkg/pipeline/handler.go:12]: handler

File: pkg/pipeline/handler.go [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+}
+
+// HandlerFunc first chan represent input
+type HandlerFunc func(context.Context, chan *Request, chan *Request) error
+
+func (si *StepInfra) Exec(ctx context.Context, cancel context.CancelFunc, p *Pipeline, input ...chan *Request) {
+	var wg sync.WaitGroup
+	// suc: rev output, in: merge request
+	in := make(chan *Request, 10)
```
Comment (gaius-qi) [inline pkg/pipeline/handler.go:21]: Use constants to record `10`.

**Review by gaius-qi — CHANGES_REQUESTED** [review]

**Pull request merged**